xtimer: Add support for arbitrary frequency hardware timers (32768 Hz support)

gebart opened the pull request:
~~Depends on #4183~~

Set XTIMER_HZ in periph_conf.h to specify the xtimer hardware timer
frequency.

The xtimer system uses the hardware timer's own tick representation
internally but a thin wrapper layer is added as a shim between xtimer
core and the public xtimer API which is based around microseconds as
the time unit.

Initially supports only 32768 Hz, and 1000000 Hz (and other power-of-two multiples of 15625), but can easily be expanded by adding a new case for XTIMER_HZ inside xtimer.h

Note: Using XTIMER_HZ = 1000000 will not add any overhead as the conversion shim will be simply an inlined straight passthrough.

**TODO**:
 - Solve how to handle time `t > 16383999997` for XTIMER_HZ == 32768
 - Add settings for cc2538 (XTIMER_HZ 16000000)
 - Test on multiple platforms and with non-standard XTIMER_HZ settings

Test subjects of particular interest:
   - atmega2560 (XTIMER_HZ == 250000)
   - cc2538 (XTIMER_HZ == 16000000)
   - mulle #4065 (XTIMER_HZ == 32768)

gebart commented:
(Renamed the include guard because it was so close to XTIMER_HZ in name)

gebart commented:
I will create a separate PR which makes use of this on Kinetis to run xtimer on the LPTMR hardware instead of PIT.

kaspar030 commented on sys/include/xtimer.h:
```diff
@@ -389,6 +393,75 @@ int xtimer_msg_receive_timeout64(msg_t *msg, uint64_t us);
+     * slow, and uses an additional 200 bytes of ROM compared to the helper
+     * function required for multiplying two 32 bit integers into a 64 bit result.
+     */
+    return ((uint64_t)(us << 9) * 0x431bde83ul) >> (12 + 32);
```
```((uint64_t)(us) * 0x431bde83ul) >> (12 + 32 - 9)``` should be equivalent, but allow the division for all values of us.

gebart replied:
Are you sure? This kind of optimization is new to me, so I don't know.

I am trying to reduce the rounding/truncation in the least significant bits by performing the multiplication first and then down-shifting after the division.
Compare `a = (x * y) / z` and `a = (x / z) * y` (using integer division)

kaspar030 replied:
Well, ```x * (whatever) / x == x/x * (whatever) == whatever```. So ```(us * 2^9 * X) / (2^44) == (2^9 * (us * X)) / (2^35)*(2^9) == (2^9)/2^9) * ((us * X) / 2^35) == (us * X) / 2^35```.

Left-shifting first might cut some bits off if ```us > 2^(32-9)```.

(I need more coffee).

gebart replied:
I was worrying about the big scary modular arithmetic magic constant would mess up the division if didn't shift first to get the proper fixed-point number.
I will update the PR.

kaspar030 replied:
wait a minute, I might have something simpler...

gebart replied:
I'm sitting on pins and needles. What's your simpler approach?

kaspar030 replied:
see #4002.

It is simpler, but it will execute the conversion macro whenever _xtimer_now() is called, e.g., a lot when spinning.

gebart replied:
Executing the conversion as sparsely as possible was the main reason I chose to put it inside the public functions and make xtimer use ticks internally. 

Also, I wanted to keep the API design around microseconds to avoid conversion macros inside the calls similar to what the hwtimer API used to require before. In my opinion it is good to have a standard unit defined for the timer delays, especially when writing portable code between platforms. 

Both the macro approach and the static inline function approach lets the compiler optimize away unnecessary conversions by precomputing the result when using constants for offsets (useful in device drivers which require delays which are specified in the data sheet of the device).

kaspar030 replied:
> Executing the conversion as sparsely as possible was the main reason I chose to put it inside the public functions and make xtimer use ticks internally.

The old vtimer used ticks internally and converted in the public functions. I specifically designed xtimer to avoid that, as it IMHO made the whole thing a lot more complex.

Not sure what is the way to go here. 

As I usually like my own solutions better, I probably can't be objective here, but I prefer doing the conversion at the latest point for several reasons:

- the xtimer code only deals with the conversions at two (very low-level) points
- all assumptions on when a timer goes into the overflow list, the long-timer list etc. stay the same across all timers
- the values for target and long_target in xtimer_t can be used without conversion
- it has slightly less code size

The downside is that the *_spin() functions are less efficient and less accurate.

kaspar030 replied:
Also, at some point we'll probably use a low-power timer as fallback, then we'd have to deal with three different ticks. Having microseconds as common base seems easier.

gebart commented on sys/include/xtimer.h:
```diff
@@ -389,6 +393,75 @@ int xtimer_msg_receive_timeout64(msg_t *msg, uint64_t us);
+    /* Saves around 1400 bytes of ROM on Cortex-M platforms (both ARMv6 and
+     * ARMv7) from avoiding linking against __aeabi_uldivmod and related helpers
+     */
+    return ((uint64_t)(us << 9) * 0x431bde82d7b634dbull) >> 12;
```
@kaspar030 is the same shift change applicable here or will it mess up anything?

authmillenon commented on sys/include/xtimer.h:
```diff
@@ -389,6 +393,75 @@ int xtimer_msg_receive_timeout64(msg_t *msg, uint64_t us);
+inline static uint32_t _xtimer_ticks_to_us(uint32_t ticks) {
+    /* Using 64 bit multiplication to avoid truncating the top 9 bits */
+    uint64_t us = (uint64_t)ticks * 15625ul;
+    return us >> 9;
```
For `_xtimer_us_to_ticks()` you commented, what the actual meaning of the calculation is. Can you please do so here, too?

gebart replied:
Right-o, will update

gebart commented:
Rebased after #3998 was merged

gebart commented on sys/include/xtimer.h:
```diff
@@ -457,49 +550,52 @@ static inline void xtimer_spin_until(uint32_t value);
-    while (_xtimer_now() > value);
-    while (_xtimer_now() < value);
+    while (_xtimer_now_ticks() > value);
+    while (_xtimer_now_ticks() < value);
```
note to self: this should be changed to `_lltimer_now_ticks`

gebart commented on sys/include/xtimer.h:
```diff
@@ -457,49 +550,52 @@ static inline void xtimer_spin_until(uint32_t value);
+
+static inline void _xtimer_spin_ticks(uint32_t offset) {
+    uint32_t start = _xtimer_now_ticks();
+    while ((_xtimer_now_ticks() - start) < offset);
```
note to self: `_lltimer_now_ticks`

kaspar030 commented:
If I define XTIMER_HZ to 32768 in cpu/native/include/periph_conf.h, examples/timer_wakeup_periodic breaks (works correctly for some a second and then loops). But that is probably a matter of correct tuning values.

gebart commented:
@kaspar030 did it run really fast?
I did not yet update the timer_init call in xtimer, so that might be one problem. I have a PR in the works for changing the meaning of the middle argument to timer_init from ticks_per_us to ticks_per_sec to handle non-"integer MHz" timers

kaspar030 commented:
> did it run really fast?
(well, I just used the 1mhz timer as 32khz timer, so I was expecting it to be ~30 times as fast)

It went as expected for maybe 10 too fast seconds, then looped. That happens if one xtimer_usleep_until() call returns too fast.

> I have a PR in the works for changing the meaning of the middle argument to timer_init from ticks_per_us to ticks_per_sec to handle non-"integer MHz" timers

nice!

gebart commented:
@kaspar030 could it be the printf's inside the test taking too much time when trying to run the main test loop at 30x speed?

kaspar030 commented:
@gebart I don't think so, as this is on native, so prints should be very fast. Also #4002 works fine at the same speed.

gebart commented:
I'll look into it.

gebart commented:
@kaspar030
>> Executing the conversion as sparsely as possible was the main reason I chose to put it inside the public functions and make xtimer use ticks internally.
>
>The old vtimer used ticks internally and converted in the public functions. I specifically designed xtimer to avoid that, as it IMHO made the whole thing a lot more complex.
>
>Not sure what is the way to go here. 
>
>As I usually like my own solutions better, I probably can't be objective here, but I prefer doing the conversion at the latest point for several reasons:
>
>- the xtimer code only deals with the conversions at two (very low-level) points
>- all assumptions on when a timer goes into the overflow list, the long-timer list etc. stay the same across all timers
>- the values for target and long_target in xtimer_t can be used without conversion
>- it has slightly less code size
>
>The downside is that the *_spin() functions are less efficient and less accurate.

You make valid points, I agree it is better to have conversions happen at the fewest possible places (fewer points of failure). The alternative method looks smaller in the example PR, but it will still need a bunch of conversions distributed across all users of the xtimer API, instead of encapsulated inside the xtimer code.

While the assumptions on which point in time (at x microseconds) the timer goes into overflow remains the same, that won't happen on the same place in the hardware counter of the different hardware timers, I think makes more sense to refer to timer _tick_ 0xffff (or 0xffffffff or whatever depending on resolution) as the last tick of the short period rather than trying to shoehorn a timer into overflowing at 65.535 ms or 16.777216 seconds or 4300-ish seconds. Debugging timers is always hard because you always affect them when trying to observe them because of the timing involved, but I think using the native timer overflow as the basis of the "long term" is the way to go here.

Regarding the values in target and long_target, I don't know, it is not a very common use case (from looking through the RIOT tree) but maybe we should provide translation functions (some kind of getter) for reading the target timer values from a struct xtimer.

The code size can probably be reduced in this PR too, I wanted something that works and is stable before messing around with optimizations.

A big difference between xtimer and vtimer is that xtimer has a better separation between xtimer public API and xtimer core. In my opinion it makes more sense to let xtimer core work with timer values which are native to the timer hardware. xtimer is the public API which should be consistent between platforms, it is quite easy to forget to use `XTIMER_USEC_TO_TICKS()` on some call (and easy to miss when reviewing).

gebart commented:
something is seriously messed up with xtimer_usleep_until in the current state of this PR. I will update the PR later.

gebart commented:
Waiting for #4040

gebart commented:
- Rebased
 - Updated to get xtimer_usleep_until working properly on native again
 - Eliminate XTIMER_SHIFT
 - Added warning for XTIMER_BACKOFF < XTIMER_OVERHEAD which causes timer underrun for small offsets
 - Added some statistics in tests/xtimer_usleep_until

gebart commented:
Rebased on top of #4040

gebart commented:
Updated to incorporate the xtimer bug fix in #4061

gebart commented:
`_xtimer_us_to_ticks64` is broken and needs updating. There should be a 64 bit right shift in there which requires 128 bit hardware integers, would work on x86_64 (with added _uint128 and >> 64), but not on any of the RIOT platforms.

hexluthor commented:
:+1: for this PR. On the `cc2538` the timers don't support a prescaler in 32-bit mode so they are stuck at 16 MHz. Would love to be able to simply `#define XTIMER_HZ 16000000`.

gebart commented:
Updated.

@hexluthor could you try the updated XTIMER_SHIFT handling?
You need to set XTIMER_HZ and XTIMER_SHIFT in board.h for your board (I don't know which board you were referring to for the 16 MHz timer)

See boards/arduino-mega2560/include/board.h for an example usage.

gebart commented:
xtimer_usleep_until has problems with the overflow at 4295 seconds. I am working on a revised version.

gebart commented:
I decided to take this in steps, first some minor cleanups and fixes in #4181, #4183 
I'm putting this PR on hold until #4040 and the smaller xtimer fixes have been merged.

gebart commented:
rebased, depends on #4183

gebart commented:
Still problems with the usec overflow at 4300-ish seconds when using a hardware timer @ 32768 Hz

gebart commented:
- Fixed the overflow problems.
 - Refactored `xtimer_usleep_until` once more.

Currently running a test on xtimer_drift to test the stability.

gebart commented:
Updated TODO in the top post.

gebart commented:
Refactored the division functions to work for very large numbers (up to 2**64)

gebart commented:
Even though the scaling for 32768 Hz works past 16000000000 usec, there's still a problem with the long timer tick at 36 something hours. Using 64 bit counts throughout can possibly fix the issue, still investigating.

gebart commented:
This PR is getting out of hand again. I will try to fork off the general improvements

OlegHahm commented:
Won't make it until the release, right?

gebart commented:
Won't make it. Will likely result in a replacement PR later.
On Mar 23, 2016 11:59 PM, "Oleg Hahm" <notifications@github.com> wrote:

> Won't make it until the release, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3990#issuecomment-200575310>
>

gebart commented:
obsoleted in favour of #5608.
